README.md
![SendGrid Logo](https://uiux.s3.amazonaws.com/2016-logos/email-logo%402x.png)

[![Build Status](https://travis-ci.org/sendgrid/smtpapi-ruby.svg?branch=master)](https://travis-ci.org/SendGrid/smtpapi-ruby)
[![Gem Version](https://badge.fury.io/rb/smtpapi.svg)](https://badge.fury.io/rb/smtpapi)
[![Email Notifications Badge](https://dx.sendgrid.com/badge/ruby)](https://dx.sendgrid.com/newsletter/ruby)
[![Gem Version](https://badge.fury.io/rb/sendgrid-ruby.svg)](https://badge.fury.io/rb/sendgrid-ruby)
[![MIT licensed](https://img.shields.io/badge/license-MIT-blue.svg)](./LICENSE.md)
[![Twitter Follow](https://img.shields.io/twitter/follow/sendgrid.svg?style=social&label=Follow)](https://twitter.com/sendgrid)
[![GitHub contributors](https://img.shields.io/github/contributors/sendgrid/smtpapi-ruby.svg)](https://github.com/sendgrid/smtpapi-ruby/graphs/contributors)

**This ruby gem allows you to quickly and more easily generate SendGrid X-SMTPAPI headers.**

# Announcements
**The default branch name for this repository has been changed to `main` as of 07/27/2020.**

All updates to this library is documented in our [CHANGELOG](https://github.com/sendgrid/smtpapi-ruby/blob/master/CHANGELOG.md).

# Table of Contents
- [Installation](#installation)
- [Quick Start](#quick-start)
- [Usage](#usage)
- [Roadmap](#roadmap)
- [How to Contribute](#contribute)
- [About](#about)
- [License](#license)

<a name="installation"></a>
# Installation

## Prerequisites

- Ruby version >= 2.4
- The SendGrid service, starting at the [free level](https://sendgrid.com/free?source=smtpapi-ruby)

## Install Package

Add this line to your application's Gemfile:

```bash
gem 'smtpapi'
```

And then execute:

```bash
bundle
```

Or install it yourself as:

```bash
gem install smtpapi
```

## Setup Environment Variables

### Environment Variable

Update the development environment with your [SENDGRID_API_KEY](https://app.sendgrid.com/settings/api_keys), for example:

```bash
echo "export SENDGRID_API_KEY='YOUR_API_KEY'" > sendgrid.env
echo "sendgrid.env" >> .gitignore
source ./sendgrid.env
```

<a name="quick-start"></a>
# Quick Start

```ruby
header = Smtpapi::Header.new
header.add_to('test@example.com')
print header.to_json
```

<a name="usage"></a>
# Usage

- [SendGrid Docs](https://sendgrid.com/docs/API_Reference/SMTP_API/index.html)
- [Example Code](https://github.com/sendgrid/smtpapi-ruby/blob/master/examples)

<a name="roadmap"></a>
# Roadmap

If you are interested in the future direction of this project, please take a look at our [milestones](https://github.com/sendgrid/smtpapi-ruby/milestones). We would love to hear your feedback.

<a name="contribute"></a>
# How to Contribute

We encourage contribution to our libraries, please see our [CONTRIBUTING](https://github.com/sendgrid/smtpapi-ruby/blob/master/CONTRIBUTING.md) guide for details.

Quick links:

- [Feature Request](https://github.com/sendgrid/smtpapi-ruby/blob/master/CONTRIBUTING.md#feature_request)
- [Bug Reports](https://github.com/sendgrid/smtpapi-ruby/blob/master/CONTRIBUTING.md#submit_a_bug_report)
- [Improvements to the Codebase](https://github.com/sendgrid/smtpapi-ruby/blob/master/CONTRIBUTING.md#improvements_to_the_codebase)
- [Review Pull Requests](https://github.com/sendgrid/smtpapi-ruby/blob/master/CONTRIBUTING.md#code-reviews)

## Credits

This library was created by [Wataru Sato](https://github.com/awwa) and is now maintained by SendGrid.

<a name="about"></a>
# About

smtpapi-ruby is maintained and funded by Twilio SendGrid, Inc. The names and logos for smtpapi-ruby are trademarks of Twilio SendGrid, Inc.

If you need help installing or using the library, please check the [Twilio SendGrid Support Help Center](https://support.sendgrid.com).

If you've instead found a bug in the library or would like new features added, go ahead and open issues or pull requests against this repo!

<a name="license"></a>
# License

[The MIT License (MIT)](LICENSE.md)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](http://semver.org/).

[2020-07-22] Version 0.1.6
--------------------------
**Library - Docs**
- [PR #82](https://github.com/sendgrid/smtpapi-ruby/pull/82): Create usage.md file. Thanks to [@prathamsharma92](https://github.com/prathamsharma92)!
- [PR #83](https://github.com/sendgrid/smtpapi-ruby/pull/83): add code review section in contributions. Thanks to [@prathamsharma92](https://github.com/prathamsharma92)!


[2020-03-04] Version 0.1.5
--------------------------
**Library - Chore**
- [PR #95](https://github.com/sendgrid/smtpapi-ruby/pull/95): add Ruby 2.7 to Travis. Thanks to [@childish-sambino](https://github.com/childish-sambino)!


[2020-02-19] Version 0.1.4
--------------------------
**Library - Chore**
- [PR #87](https://github.com/sendgrid/smtpapi-ruby/pull/87): Add test coverage analysis. Thanks to [@RolandBurrows](https://github.com/RolandBurrows)!


[2020-01-22] Version 0.1.3
--------------------------
**Library - Fix**
- [PR #94](https://github.com/sendgrid/smtpapi-ruby/pull/94): add skip_cleanup flag to fix travis deploy. Thanks to [@thinkingserious](https://github.com/thinkingserious)!


[2020-01-22] Version 0.1.2
--------------------------
**Library - Docs**
- [PR #93](https://github.com/sendgrid/smtpapi-ruby/pull/93): baseline all the templated markdown docs. Thanks to [@childish-sambino](https://github.com/childish-sambino)!


[2020-01-09] Version 0.1.1
--------------------------
**Library - Chore**
- [PR #92](https://github.com/sendgrid/smtpapi-ruby/pull/92): Auto-deploy upon successful tagged commit. Thanks to [@thinkingserious](https://github.com/thinkingserious)!
- [PR #50](https://github.com/sendgrid/smtpapi-ruby/pull/50): Update License date range. Thanks to [@shucon](https://github.com/shucon)!
- [PR #85](https://github.com/sendgrid/smtpapi-ruby/pull/85): updated Readme.md. Thanks to [@rahulpuroht](https://github.com/rahulpuroht)!
- [PR #67](https://github.com/sendgrid/smtpapi-ruby/pull/67): Add Use Cases directory. Thanks to [@eyewritecode](https://github.com/eyewritecode)!
- [PR #65](https://github.com/sendgrid/smtpapi-ruby/pull/65): Update LICENSE.txt. Thanks to [@nocategory](https://github.com/nocategory)!
- [PR #64](https://github.com/sendgrid/smtpapi-ruby/pull/64): Resolve linting issues. Thanks to [@mrala](https://github.com/mrala)!
- [PR #46](https://github.com/sendgrid/smtpapi-ruby/pull/46): Replaced .length > 0 with .!empty?. Thanks to [@nvzard](https://github.com/nvzard)!
- [PR #35](https://github.com/sendgrid/smtpapi-ruby/pull/35): Replaced .length > 0 with !empty #32. Thanks to [@181514912](https://github.com/181514912)!
- [PR #42](https://github.com/sendgrid/smtpapi-ruby/pull/42): Added unittest to check for specific repo files. Thanks to [@mptap](https://github.com/mptap)!
- [PR #40](https://github.com/sendgrid/smtpapi-ruby/pull/40): Update LICENSE.txt. Thanks to [@GaryLeutheuser](https://github.com/GaryLeutheuser)!
- [PR #41](https://github.com/sendgrid/smtpapi-ruby/pull/41): Add License end year test. Thanks to [@diegous](https://github.com/diegous)!
- [PR #43](https://github.com/sendgrid/smtpapi-ruby/pull/43): Update LICENSE.txt. Thanks to [@astromoose](https://github.com/astromoose)!
- [PR #26](https://github.com/sendgrid/smtpapi-ruby/pull/26): add a .env_sample file. Thanks to [@thepriefy](https://github.com/thepriefy)!
- [PR #24](https://github.com/sendgrid/smtpapi-ruby/pull/24): add Gem badge to README file. Thanks to [@thepriefy](https://github.com/thepriefy)!
- [PR #28](https://github.com/sendgrid/smtpapi-ruby/pull/28): Added the gem badge. Thanks to [@roooodcastro](https://github.com/roooodcastro)!
- [PR #44](https://github.com/sendgrid/smtpapi-ruby/pull/44): Rename PULL_REQUEST_TEMPLATE.md. Thanks to [@thepriefy](https://github.com/thepriefy)!
- [PR #25](https://github.com/sendgrid/smtpapi-ruby/pull/25): add pull request template. Thanks to [@rohitdaryanani](https://github.com/rohitdaryanani)!
- [PR #31](https://github.com/sendgrid/smtpapi-ruby/pull/31): Added .codeclimate.yml. Thanks to [@veruz](https://github.com/veruz)!
- [PR #18](https://github.com/sendgrid/smtpapi-ruby/pull/18): add .github/ISSUE_TEMPLATE. Thanks to [@thepriefy](https://github.com/thepriefy)!
- [PR #13](https://github.com/sendgrid/smtpapi-ruby/pull/13): Add/update badges on README. Thanks to [@jessecalton](https://github.com/jessecalton)!
- [PR #12](https://github.com/sendgrid/smtpapi-ruby/pull/12): more SEO friendly url. Thanks to [@ladhadha](https://github.com/ladhadha)!

**Library - Fix**
- [PR #69](https://github.com/sendgrid/smtpapi-ruby/pull/69): Remove duplicate text in CONTRIBUTING.md. Thanks to [@eyewritecode](https://github.com/eyewritecode)!
- [PR #17](https://github.com/sendgrid/smtpapi-ruby/pull/17): Fixed some Grammatical errors in CONTRIBUTING.md. Thanks to [@mohsincl](https://github.com/mohsincl)!
- [PR #16](https://github.com/sendgrid/smtpapi-ruby/pull/16): update contributing and readme - fix typo and ToC. Thanks to [@pushkyn](https://github.com/pushkyn)!

**Library - Docs**
- [PR #15](https://github.com/sendgrid/smtpapi-ruby/pull/15): update README.md. Thanks to [@thepriefy](https://github.com/thepriefy)!
- [PR #14](https://github.com/sendgrid/smtpapi-ruby/pull/14): add table of contents in README.md. Thanks to [@thepriefy](https://github.com/thepriefy)!
- [PR #9](https://github.com/sendgrid/smtpapi-ruby/pull/9): Add TROUBLESHOOTING.md. Thanks to [@colto](https://github.com/colto)!
- [PR #7](https://github.com/sendgrid/smtpapi-ruby/pull/7): Added Code of Conduct. Thanks to [@gr8shivam](https://github.com/gr8shivam)!
- [PR #5](https://github.com/sendgrid/smtpapi-ruby/pull/5): Update README.md. Thanks to [@ciceropablo](https://github.com/ciceropablo)!


[2015-04-06] Version 0.1.0
--------------------------
### Changed
- Change home to SendGrid
- Add unit test dev dep


test/test.rb
require 'test_helper'
require 'test/unit'
require './lib/smtpapi'

#
# SmtpapiTest implementation
#
class SmtpapiTest < Test::Unit::TestCase
  def test_version
    assert_equal('0.1.6', Smtpapi::VERSION)
  end

  def test_empty
    header = Smtpapi::Header.new
    assert_equal('{}', header.json_string)
  end

  def test_add_to
    header = Smtpapi::Header.new
    header.add_to('you@youremail.com')
    header.add_to('other@otheremail.com', 'Other Name')
    assert_equal(
      '{"to":["you@youremail.com","Other Name <other@otheremail.com>"]}',
      header.json_string
    )
  end

  def test_add_to_array
    header = Smtpapi::Header.new
    header.add_to(['you@youremail.com', 'my@myemail.com'])
    assert_equal(
      '{"to":["you@youremail.com","my@myemail.com"]}',
      header.json_string
    )
  end

  def test_set_tos
    header = Smtpapi::Header.new
    header.set_tos(['you@youremail.com', 'other@otheremail.com'])
    assert_equal(
      '{"to":["you@youremail.com","other@otheremail.com"]}',
      header.json_string
    )
  end

  def test_add_substitution
    header = Smtpapi::Header.new
    header.add_substitution('keep', ['secret'])
    header.add_substitution('other', %w(one two))
    assert_equal(
      '{"sub":{"keep":["secret"],"other":["one","two"]}}',
      header.json_string
    )
  end

  def test_set_substitutions
    header = Smtpapi::Header.new
    header.set_substitutions('keep' => ['secret'])
    assert_equal('{"sub":{"keep":["secret"]}}', header.json_string)
  end

  def test_add_section
    header = Smtpapi::Header.new
    header.add_section('-charge-', 'This ship is useless.')
    header.add_section('-bomber-', 'Only for sad vikings.')
    assert_equal(
      '{"section":'\
        '{"-charge-":"This ship is useless.",'\
          '"-bomber-":"Only for sad vikings."}}',
      header.json_string
    )
  end

  def test_set_sections
    header = Smtpapi::Header.new
    header.set_sections('-charge-' => 'This ship is useless.')
    assert_equal(
      '{"section":{"-charge-":"This ship is useless."}}',
      header.json_string
    )
  end

  def test_add_unique_arg
    header = Smtpapi::Header.new
    header.add_unique_arg('cat', 'dogs')
    assert_equal('{"unique_args":{"cat":"dogs"}}', header.json_string)
  end

  def test_set_unique_args
    header = Smtpapi::Header.new
    header.set_unique_args('cow' => 'chicken')
    header.set_unique_args('dad' => 'proud')
    assert_equal('{"unique_args":{"dad":"proud"}}', header.json_string)
  end

  def test_add_category
    header = Smtpapi::Header.new
    header.add_category('tactics')
    header.add_category('advanced')
    assert_equal('{"category":["tactics","advanced"]}', header.json_string)
  end

  def test_set_categories
    header = Smtpapi::Header.new
    header.set_categories(%w(tactics advanced))
    assert_equal('{"category":["tactics","advanced"]}', header.json_string)
  end

  def test_add_filter
    header = Smtpapi::Header.new
    header.add_filter('footer', 'enable', 1)
    header.add_filter('footer', 'text/html', '<strong>boo</strong>')
    assert_equal(
      '{"filters":'\
        '{"footer":'\
          '{"settings":'\
            '{"enable":1,'\
              '"text/html":"<strong>boo</strong>"'\
            '}'\
          '}'\
        '}'\
      '}',
      header.json_string
    )
  end

  def test_set_filters
    header = Smtpapi::Header.new
    filter = {
      'footer' => {
        'setting' => {
          'enable' => 1,
          'text/plain' => 'You can haz footers!'
        }
      }
    }
    header.set_filters(filter)
    assert_equal(
      '{"filters":'\
        '{"footer":'\
          '{"setting":'\
            '{"enable":1,"text/plain":"You can haz footers!"}'\
          '}'\
        '}'\
      '}',
      header.json_string
    )
  end

  def test_add_category_unicode
    header = Smtpapi::Header.new
    header.add_category('Martí')
    header.add_category('天破活殺')
    header.add_category('天翔十字鳳')
    assert_equal(
      '{"category":'\
        '["Mart\\u00ed",'\
          '"\\u5929\\u7834\\u6d3b\\u6bba",'\
          '"\\u5929\\u7fd4\\u5341\\u5b57\\u9cf3"]}',
      header.json_string
    )
    header.add_category('鼖')
    assert_equal(
      '{"category":'\
        '["Mart\\u00ed",'\
          '"\\u5929\\u7834\\u6d3b\\u6bba",'\
          '"\\u5929\\u7fd4\\u5341\\u5b57\\u9cf3",'\
          '"\\ud87e\\ude1b"]}',
      header.json_string
    )
  end

  def test_sent_send_at
    header = Smtpapi::Header.new
    localtime = Time.local(2014, 8, 29, 17, 56, 35)
    header.set_send_at(localtime)

    assert_equal("{\"send_at\":#{localtime.to_i}}", header.json_string)
  end

  def test_send_each_at
    header = Smtpapi::Header.new
    localtime1 = Time.local(2014, 8, 29, 17, 56, 35)
    localtime2 = Time.local(2013, 12, 31, 0, 0, 0)
    localtime3 = Time.local(2015, 9, 1, 4, 5, 6)
    header.set_send_each_at([localtime1, localtime2, localtime3])

    assert_equal(
      '{"send_each_at":'\
        "[#{localtime1.to_i},#{localtime2.to_i},#{localtime3.to_i}]"\
      '}',
      header.json_string
    )
  end

  def test_asm_group_id
    header = Smtpapi::Header.new
    header.set_asm_group(2)

    assert_equal('{"asm_group_id":2}', header.json_string)
  end

  def test_ip_pool
    header = Smtpapi::Header.new
    header.set_ip_pool('test_pool')

    assert_equal('{"ip_pool":"test_pool"}', header.json_string)
  end

  # def test_docker_exists
  #   assert(File.file?('./Dockerfile') || File.file?('./docker/Dockerfile'))
  # end

  # def test_docker_compose_exists
  #   assert(
  #     File.file?('./docker-compose.yml') ||
  #     File.file?('./docker/docker-compose.yml')
  #   )
  # end

  def test_env_sample_exists
    assert(File.file?('./.env_sample'))
  end

  def test_gitignore_exists
    assert(File.file?('./.gitignore'))
  end

  def test_travis_exists
    assert(File.file?('./.travis.yml'))
  end

  def test_codeclimate_exists
    assert(File.file?('./.codeclimate.yml'))
  end

  def test_changelog_exists
    assert(File.file?('./CHANGELOG.md'))
  end

  def test_code_of_conduct_exists
    assert(File.file?('./CODE_OF_CONDUCT.md'))
  end

  def test_contributing_exists
    assert(File.file?('./CONTRIBUTING.md'))
  end

  def test_issue_template_exists
    assert(File.file?('./ISSUE_TEMPLATE.md'))
  end

  def test_license_exists
    assert(File.file?('./LICENSE.md') || File.file?('./LICENSE.txt'))
  end

  def test_pull_request_template_exists
    assert(File.file?('./PULL_REQUEST_TEMPLATE.md'))
  end

  def test_readme_exists
    assert(File.file?('./README.md'))
  end

  def test_troubleshooting_exists
    assert(File.file?('./TROUBLESHOOTING.md'))
  end

  # def test_usage_exists
  #   assert(File.file?('./USAGE.md'))
  # end

  # def test_use_cases_exists
  #   assert(File.file?('./USE_CASES.md'))
  # end

  def test_license_date_is_updated
    license_year = IO.read('LICENSE.md').match(
      /Copyright \(C\) (\d{4}), Twilio SendGrid/
    )[1]
    current_year = Time.new.year

    assert_equal(current_year, license_year.to_i)
  end
end


lib/smtpapi.rb
$LOAD_PATH.unshift File.dirname(__FILE__)
require 'smtpapi/version'
require 'json'

module Smtpapi
  #
  # SendGrid smtpapi header implementation
  #
  class Header
    attr_reader :to, :sub, :section, :category, :unique_args, :filters
    attr_reader :send_at, :send_each_at, :asm_group_id, :ip_pool

    def initialize
      @to = []
      @sub = {}
      @section = {}
      @category = []
      @unique_args = {}
      @filters = {}
      @send_at = nil
      @send_each_at = []
      @asm_group_id = nil
      @ip_pool = nil
    end

    def add_to(address, name = nil)
      if address.is_a?(Array)
        @to.concat(address)
      else
        value = address
        value = "#{name} <#{address}>" unless name.nil?
        @to.push(value)
      end
      self
    end

    def set_tos(addresses)
      @to = addresses
      self
    end

    def add_substitution(sub, values)
      @sub[sub] = values
      self
    end

    def set_substitutions(key_value_pairs)
      @sub = key_value_pairs
      self
    end

    def add_section(key, value)
      @section[key] = value
      self
    end

    def set_sections(key_value_pairs)
      @section = key_value_pairs
      self
    end

    def add_unique_arg(key, value)
      @unique_args[key] = value
      self
    end

    def set_unique_args(key_value_pairs)
      @unique_args = key_value_pairs
      self
    end

    def add_category(category)
      @category.push(category)
      self
    end

    def set_categories(categories)
      @category = categories
      self
    end

    def add_filter(filter_name, parameter_name, parameter_value)
      @filters[filter_name] = {} if @filters[filter_name].nil?
      @filters[filter_name]['settings'] = {} if @filters[filter_name]['settings'].nil?
      @filters[filter_name]['settings'][parameter_name] = parameter_value
      self
    end

    def set_filters(filters)
      @filters = filters
      self
    end

    def set_send_at(send_at)
      @send_at = send_at
      self
    end

    def set_send_each_at(send_each_at)
      @send_each_at = send_each_at
      self
    end

    def set_asm_group(group_id)
      @asm_group_id = group_id
      self
    end

    def set_ip_pool(pool_name)
      @ip_pool = pool_name
      self
    end

    def to_array
      data = {}
      data['to'] = @to unless @to.empty?
      data['sub'] = @sub unless @sub.empty?
      data['section'] = @section unless @section.empty?
      data['unique_args'] = @unique_args unless @unique_args.empty?
      data['category'] = @category unless @category.empty?
      data['filters'] = @filters unless @filters.empty?
      data['send_at'] = @send_at.to_i unless @send_at.nil?
      data['asm_group_id'] = @asm_group_id.to_i unless @asm_group_id.nil?
      data['ip_pool'] = @ip_pool unless @ip_pool.nil?
      str_each_at = []

      @send_each_at.each do |val|
        str_each_at.push(val.to_i)
      end

      data['send_each_at'] = str_each_at unless str_each_at.empty?

      data
    end

    protected :to_array

    def json_string
      escape_unicode(to_array.to_json)
    end

    alias :to_json :json_string

    def escape_unicode(str)
      str.unpack('U*').map do |i|
        if i > 65_535
          "\\u#{format('%04x', ((i - 0x10000) / 0x400 + 0xD800))}" \
          "\\u#{format('%04x', ((i - 0x10000) % 0x400 + 0xDC00))}"
        elsif i > 127
          "\\u#{format('%04x', i)}"
        else
          i.chr('UTF-8')
        end
      end.join
    end
  end
end


examples/example.rb
require_relative '../lib/smtpapi'
header = Smtpapi::Header.new

## add_to
header.add_to('test1@example.com')
header.add_to('test2@example.com', 'other')
# header.add_to(['test1@example.com', 'test2@example.com'])

## set_tos
# header.set_tos(['you@youremail.com', 'other@otheremail.com'])
# to => ['you@youremail.com', 'other@otheremail.com']

## add_substitution
# sub = {keep: ['secret']}
header.add_substitution('keep', ['secret'])
# sub = {keep: ['secret'], other: ['one', 'two']}
header.add_substitution('other', %w(one two))

## set_substitutions
# header.set_substitutions({'keep' => 'secret'})  # sub = {keep: ['secret']}

## add_unique_arg
header.add_unique_arg('cat', 'dogs')

## set_unique_args
# header.set_unique_args({'cow' => 'chicken'})
# header.set_unique_args({'dad' => 'proud'})

## add_category
# category = ['tactics']
header.add_category('tactics')
# category = ['tactics', 'advanced']
header.add_category('advanced')

## set_categories
# category = ['tactics', 'advanced']
# header.set_categories(['tactics', 'advanced'])

## add_section
header.add_section('-charge-', 'This ship is useless.')
header.add_section('-bomber-', 'Only for sad vikings.')

## set_sections
# header.set_sections({'-charge-' => 'This ship is useless.'})

## add_filter
header.add_filter('footer', 'enable', 1)
header.add_filter('footer', 'text/html', '<strong>boo</strong>')

## set_filters
# filter = {
#   'footer' => {
#     'setting' => {
#       'enable' => 1,
#       "text/plain" => 'You can haz footers!'
#     }
#   }
# }
# header.set_filters(filter)

## set_send_at
lt = Time.local(2014, 8, 29, 17, 56, 35)
header.set_send_at(lt)

## set_send_each_at
# lt1       = Time.local(2014,  8, 29, 17, 56, 35)
# lt2       = Time.local(2013, 12, 31,  0,  0,  0)
# lt3       = Time.local(2015,  9,  1,  4,  5,  6)
# header.set_send_each_at([lt1, lt2, lt3])

## asm_group_id
# This is to specify an ASM Group for the message.
# See: https://sendgrid.com/docs/User_Guide/advanced_suppression_manager.html
header.set_asm_group(2)

## set_ip_pool
# Using IP Pools with the SMTP API Header
# See: https://sendgrid.com/docs/API_Reference/Web_API_v3/IP_Management/ip_pools.html)
header.set_ip_pool('test_pool')

print header.to_json
